src/v/cluster/decommissioning_monitor.cc

```diff
@@ -0,0 +1,256 @@
+    // ask partition allocator for reallocation of
+    // replicas
+    vlog(clusterlog.debug, "reallocating partition {}", meta.ntp);
+    auto f = ss::now();
+    if (!meta.new_assignment) {
+        f = _allocator
+              .invoke_on(
+                partition_allocator::shard,
+                [current = meta.current_assignment](
+                  partition_allocator& allocator) mutable {
+                    return allocator.reallocate_decommissioned_replicas(
+                      current);
+                })
+              .then([&meta](
+                      std::optional<partition_allocator::allocation_units> u) {
+                  meta.new_assignment = std::move(u);
+              });
+    }
+
+    return f.then([this, &meta] {
+        if (!meta.new_assignment) {
+            return ss::now();
+        }
+        vlog(
+          clusterlog.debug,
+          "moving replicas of {} to {}",
+          meta.ntp,
+          meta.new_assignment->get_assignments().begin()->replicas);
+        return _topics_frontend.local()
+          .move_partition_replicas(
+            meta.ntp,
+            meta.new_assignment->get_assignments().begin()->replicas,
+            model::timeout_clock::now() + _retry_timeout)
+          .then([&meta, this](std::error_code ec) {
+              vlog(
+                clusterlog.debug, "moving replicas result: {}", ec.message());
+              auto it = _topics.local().topics_map().find(
+                model::topic_namespace_view(meta.ntp));
+              // topic was deleted, we are done with assignment
+              if (it == _topics.local().topics_map().end()) {
+                  meta.is_finished = true;
+                  return;
+              }
+              meta.is_finished = is_replica_set_up_to_date(
+                it->second,
+                meta.ntp,
+                meta.new_assignment->get_assignments().begin()->replicas);
+          });
+    });
+}
+
+bool is_in_replica_set(
+  const std::vector<model::broker_shard>& replicas,
+  const absl::flat_hash_set<model::node_id>& ids) {
+    auto it = std::find_if(
+      std::cbegin(replicas),
+      std::cend(replicas),
+      [&ids](const model::broker_shard& bs) {
+          return ids.contains(bs.node_id);
+      });
+
+    return it != std::cend(replicas);
+}
+
+absl::flat_hash_set<model::node_id> ids_set(std::vector<model::node_id> ids) {
+    absl::flat_hash_set<model::node_id> res;
+    res.reserve(ids.size());
+    for (auto& id : ids) {
+        res.emplace(id);
+    }
+
+    return res;
+}
+bool decommissioning_monitor::is_decommissioned(model::node_id id) const {
+    return std::find(
+             std::cbegin(_decommissioned), std::cend(_decommissioned), id)
+           != _decommissioned.cend();
+}
+
+void decommissioning_monitor::decommission(std::vector<model::node_id> ids) {
+    absl::flat_hash_set<model::node_id> set_ids;
+    for (auto id : ids) {
```

Comment:
above you use `    for (auto& id : ids) {` is this a heavy weight thing so we want to avoid the copy?